subprojects/core/src/main/java/org/gradle/api/provider/ConfigurablePropertyState.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.provider;
+
+import org.gradle.api.Incubating;
+
+import java.util.Set;
+
+/**
+ * A mutable {@code PropertyState} representation.
+ *
+ * @param <T> Type of value represented by property state
+ * @since 3.5
+ */
+@Incubating
+public interface ConfigurablePropertyState<T> extends PropertyState<T> {
```

Comment:
This should be `ConfigurableProvider<T> extends Provider<T>`. This type is used to represent ad hoc providers that aren't the property of something.